lib/spack/spack/cmd/buildcache.py

```diff
@@ -320,13 +320,20 @@ def _createtarball(env, spec_yaml, packages, directory, key, no_deps, force,
     pkgs = set(packages)
     specs = set()
 
-    outdir = '.'
-    if directory:
+    if not directory:
+        msg = 'Specify directory name or mirror name for outputing buildcaches'
+        msg += ' to with -d option. Use \n'
+        msg += 'spack buildache create -d <dir name|mirror name> ...'
+        tty.die(msg)
+    else:
```

Comment:
I [still](https://github.com/spack/spack/pull/15090#pullrequestreview-362283865) think just making `-d` a required argument is the way to go.  That + a better help string for the `-d` option should be enough for users to know what information they'd need to provide.

Reply:
Ok. How do I tell argparse this is a required argument?

Reply:
I added the required=True option but now the next positional argument is save into the directory variable even if it is a spec because someone forgot to add the directory or mirror after -d.

Reply:
This is especially bad because anonymous specs can be passed in as /hash. Since this could be passed to spack.mirror.MirrorCollection().lookup(outdir) can a check that the path exists be done?

Reply:
```
(flake8) [gartung@gartung spack]$ spack buildcache create -r -a -d  /3n2qkuj /skjzxw6 
==> creating binary cache file for package git@2.25.0%gcc@4.8.5~tcltk arch=linux-scientific7-x86_64 
==> Warning: Missing a source id for curl@7.29.0
==> Warning: Missing a source id for expat@2.1.0
==> Warning: Missing a source id for libxml2@2.9.1
==> Warning: Missing a source id for tar@1.26
==> Warning: Missing a source id for zlib@1.2.7
==> Warning: [Errno 13] Permission denied: '/3n2qkuj'
==> creating binary cache file for package gettext@0.20.1%gcc@4.8.5+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-scientific7-x86_64 
==> Warning: [Errno 13] Permission denied: '/3n2qkuj'
==> creating binary cache file for package libiconv@1.16%gcc@4.8.5 arch=linux-scientific7-x86_64 
==> Warning: [Errno 13] Permission denied: '/3n2qkuj'
==> creating binary cache file for package pcre2@10.31%gcc@4.8.5+multibyte arch=linux-scientific7-x86_64 
==> Warning: [Errno 13] Permission denied: '/3n2qkuj'
==> creating binary cache file for package libunistring@0.9.10%gcc@4.8.5 arch=linux-scientific7-x86_64 
==> Warning: [Errno 13] Permission denied: '/3n2qkuj'
==> creating binary cache file for package libidn2@2.1.1a%gcc@4.8.5 arch=linux-scientific7-x86_64 
==> Warning: [Errno 13] Permission denied: '/3n2qkuj'
```

Reply:
> I added the required=True option but now the next positional argument is saved into the directory variable even if it is a spec because someone forgot to add the directory or mirror after -d.

1.  To whatever extent this parsing behavior is problematic, it predates this PR, and there hasn't yet been any changes proposed here that would eliminate it.  The addition of `required=True` to `argparse` is unrelated.

2. That's exactly what we want, isn't it? The whole point of making `-d` a required argument is to eliminate a default setting for it, thereby requiring that users provide a mirror name or url as a value to the `-d` flag.  If the user doesn't do this and provides a spec that happens to also be a valid mirror URL, I'm not sure there's much we can do to detect that kind of misuse.  There are tons of examples of this same kind of ambiguity in cli arg parsing where the onus is entirely on the user to ensure that they provide correct arguments.

```
 $ echo 'Hello, World' > a
 $ echo 'Hello, my darling' > b
 $ echo '[hH]ello' > patterns
 $ grep -f patterns b c
b:Hello, World
c:Hello, my darling
 $ grep -f b c # whoops
```

> This is especially bad because anonymous specs can be passed in as /hash.

And if a user says `-d /hash` to Spack in this context, I don't think it's right to assume that they made a mistake, and didn't want to create a buildcache that lived literally under `/hash`.

> Since this could be passed to spack.mirror.MirrorCollection().lookup(outdir) can a check that the path exists be done?

Perhaps, but right now, the directory for a mirror is created (if it doesn't exist) as part of the creation process.  We could remove that behavior and require that users pre-create the directory, but then 1 - the operation is not much of a "create" at that point, and 2 - the requirement doesn't generalize to non-local mirrors.

I agree that helping users avoid mistakes makes sense, but we shouldn't go crazy trying to head off every conceivable mistake a user might make.

As to your code example, I agree that the behavior is not ideal, but I think the better way to address it is to fail fast on the permission denied errors.  It looks like part of the code is not catching those errors and it should be.

Reply:
Just to be clear: I don't think my point above about catching the permission denied errors needs to be incorporated for this PR, necessarily.  Someone (probably me, since I wrote the code) can handle that in another PR.

Reply:
#15157